packages/zero-cache/src/services/change-source/custom/change-source.ts

```diff
@@ -0,0 +1,248 @@
+import {computeZqlSpecs} from '../../../db/lite-tables.ts';
+import {StatementRunner} from '../../../db/statements.ts';
+import {stringify} from '../../../types/bigint-json.ts';
+import {stream} from '../../../types/streams.ts';
+import type {
+  ChangeSource,
+  ChangeStream,
+} from '../../change-streamer/change-streamer-service.ts';
+import {type ReplicationConfig} from '../../change-streamer/schema/tables.ts';
+import {ChangeProcessor} from '../../replicator/change-processor.ts';
+import {initChangeLog} from '../../replicator/schema/change-log.ts';
+import {
+  getSubscriptionState,
+  initReplicationState,
+} from '../../replicator/schema/replication-state.ts';
+import type {ShardConfig} from '../pg/shard-config.ts';
+import {changeStreamMessageSchema} from '../protocol/current/downstream.ts';
+import {type ChangeSourceUpstream} from '../protocol/current/upstream.ts';
+import {initSyncSchema} from './sync-schema.ts';
+
+/**
+ * Initializes a Postgres change source, including the initial sync of the
+ * replica, before streaming changes from the corresponding logical replication
+ * stream.
+ */
+export async function initializeCustomChangeSource(
+  lc: LogContext,
+  upstreamURI: string,
+  shard: ShardConfig,
+  replicaDbFile: string,
+): Promise<{replicationConfig: ReplicationConfig; changeSource: ChangeSource}> {
+  await initSyncSchema(
+    lc,
+    `replica-${shard.id}`,
+    shard,
+    replicaDbFile,
+    upstreamURI,
+  );
+
+  const replica = new Database(lc, replicaDbFile);
+  const replicationConfig = getSubscriptionState(new StatementRunner(replica));
+  replica.close();
+
+  if (shard.publications.length) {
+    // Verify that the publications match what has been synced.
+    const requested = [...shard.publications].sort();
+    const replicated = replicationConfig.publications.sort();
+    if (!deepEqual(requested, replicated)) {
+      throw new Error(
+        `Invalid ShardConfig. Requested publications [${requested}] do not match synced publications: [${replicated}]`,
+      );
+    }
+  }
+
+  const changeSource = new CustomChangeSource(
+    lc,
+    upstreamURI,
+    shard.id,
+    replicationConfig,
+  );
+
+  return {replicationConfig, changeSource};
+}
+
+class CustomChangeSource implements ChangeSource {
+  readonly #lc: LogContext;
+  readonly #upstreamUri: string;
+  readonly #shardID: string;
+  readonly #replicationConfig: ReplicationConfig;
+
+  constructor(
+    lc: LogContext,
+    upstreamUri: string,
+    shardID: string,
+    replicationConfig: ReplicationConfig,
+  ) {
+    this.#lc = lc.withContext('component', 'change-source');
+    this.#upstreamUri = upstreamUri;
+    this.#shardID = shardID;
+    this.#replicationConfig = replicationConfig;
+  }
+
+  initialSync(): ChangeStream {
+    return this.#startStream();
+  }
+
+  startStream(clientWatermark: string): Promise<ChangeStream> {
+    return Promise.resolve(this.#startStream(clientWatermark));
```

Comment:
Can `#startStream` throw? Maybe use async function to make this return a promise rejection in case `#startStream` throws.